Add SqlServer MSI Support

samfoxcode opened the pull request:
Add the option for MSI Auth to Azure SQL Server. Expects a param to be passed to enable MSI Auth, useMsi=true.

- update go-mssqldb module to pull in msi auth support
- make code change in database/sqlserver/sqlserver.go to check for useMsi param
- dynamically get database resource uri based on the server uri
- adds [adal](https://pkg.go.dev/github.com/Azure/go-autorest/autorest/adal) module which is needed for fetching the msi token

PR from [denisenkom/go-mssqldb](https://github.com/denisenkom/go-mssqldb) that added msi support: [#546](https://github.com/denisenkom/go-mssqldb/pull/546)

samfoxcode committed:
MSI Auth for SQL Server

Add MSI Auth option to SQL Server connection. Default if no password is provided in connection string.

samfoxcode committed:
Parse resource endpoint from server for msi

update host name parsing to get just the resource endpoint for msi

samfoxcode committed:
Update go-mssqldb

update go-mssqldb to resolve panic issue referenced here: https://github.com/denisenkom/go-mssqldb/pull/642

samfoxcode committed:
Update sqlserver.go

switch from deprecated methods. NewServicePrincipalTokenFromManagedIdentity calls the two methods that are deprecated for us

samfoxcode committed:
Update sqlserver.go

add useMsi param instead of looking for nil password

samfoxcode committed:
Update sqlserver readme

Update sqlserver readme for msi auth. make useMsi a bit safer

samfoxcode committed:
Update sqlserver.go

remove comment

kfcampbell commented on database/sqlserver/sqlserver.go:
```diff
@@ -116,16 +119,43 @@ func WithInstance(instance *sql.DB, config *Config) (database.Driver, error) {
+
+	var db *sql.DB
+	if useMsi {
+		tokenProvider, err := getMSITokenProvider(fmt.Sprintf("%s%s", "https://", strings.Join(strings.Split(purl.Hostname(), ".")[1:], ".")))
```
Could this line be refactored into a helper in order to make its purpose more clear and maybe make it testable? I'm a little worried about some of these string operations being brittle with unexpected input.

samfoxcode replied:
yeah, I agree. The main purpose here is to get the resource uri which could change across clouds. I'll see what I can do. Currently, if the uri is incorrect, I believe it would result in a 4xx (invalid resource uri for tenant or something similar). Will validate that behavior and update

kfcampbell replied:
Sounds good, thank you! The cloud change use case is not something I had picked up from reading the code but that's super important.

kfcampbell commented on database/sqlserver/README.md:
```diff
@@ -16,6 +16,7 @@
 | `dial+timeout` | | in seconds (default is 15), set to 0 for no timeout. |
 | `encrypt` | | `disable` - Data send between client and server is not encrypted. `false` - Data sent between client and server is not encrypted beyond the login packet (Default). `true` - Data sent between client and server is encrypted. |
 | `app+name` || The application name (default is go-mssqldb). |
+| `useMsi` | | `true` - Use Azure MSI Authentication for connecting to Sql Server. Must be running from an Azure VM/an instance with MSI enabled. `false` - Use password authentication (Default). Azure MSI Auth details: https://docs.microsoft.com/en-us/azure/app-service/app-service-web-tutorial-connect-msi.
```
```suggestion
| `useMsi` | | `true` - Use Azure MSI Authentication for connecting to Sql Server. Must be running from an Azure VM/an instance with MSI enabled. `false` - Use password authentication (Default). See [here for Azure MSI Auth details](https://docs.microsoft.com/en-us/azure/app-service/app-service-web-tutorial-connect-msi).
```

kfcampbell commented on database/sqlserver/sqlserver.go:
```diff
@@ -116,16 +119,43 @@ func WithInstance(instance *sql.DB, config *Config) (database.Driver, error) {
+	var db *sql.DB
+	if useMsi {
+		tokenProvider, err := getMSITokenProvider(fmt.Sprintf("%s%s", "https://", strings.Join(strings.Split(purl.Hostname(), ".")[1:], ".")))
+		if err != nil {
```
Out of curiosity, what happens if I pass `useMsi` as `true` but there's no provider available? Is it a timeout or an immediate failure?

samfoxcode replied:
if the resource uri is invalid it'll timeout, but if there isn't an identity available it seems to just hang - looking into why

kfcampbell replied:
Sounds good! I'm trying to think like the maintainer of the upstream package might and try to identify any failure modes introduced. I don't think there's necessarily a specific action item here and we shouldn't block merging on this comment, but I am certainly curious.

kfcampbell commented on database/sqlserver/sqlserver.go:
```diff
@@ -116,16 +119,43 @@ func WithInstance(instance *sql.DB, config *Config) (database.Driver, error) {
+			return nil, err
+		}
+
+		db = sql.OpenDB(connector)
```
What's the difference between this `sql.OpenDB()` call and the below `sql.Open()` call? What's the reason for using two different methods to open the database?

samfoxcode replied:
sql.Open will actually end up calling sql.OpenDB behind the scenes, but it builds the connector first based on what's passed in/the db context. sql.OpenDB just expects a connector, which is this case we're building ourselves, so we can call sql.OpenDB directly in that case. This follows the mssql guidance for using msi connections: [mssql msi](https://github.com/denisenkom/go-mssqldb#azure-active-directory-authentication---preview)

source code which has sql.Open and sql.OpenDB (sorry, can't permalink to sql.Open): https://golang.org/src/database/sql/sql.go

kfcampbell replied:
Awesome, thanks for the explanation!

kfcampbell commented on database/sqlserver/sqlserver.go:
```diff
@@ -344,3 +374,16 @@ func (ss *SQLServer) ensureVersionTable() (err error) {
 	return nil
 }
+
+func getMSITokenProvider(resource string) (func() (string, error), error) {
```
Do you mind briefly walking me through the decision to use a higher order function here? What is it buying us?

kfcampbell replied:
Ohh, I see in [this PR](https://github.com/denisenkom/go-mssqldb/pull/546/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R125) it's due to the short-lived nature of AAD tokens and this function ensures a refresh.

samfoxcode replied:
mssql.NewAccessTokenConnector expects a callback which is used to [build the connection](https://github.com/denisenkom/go-mssqldb/blob/35162394aa7974ac37c4b42ddb1c200f6b01ebac/accesstokenconnector.go#L14) and that will be used to retrieve fresh tokens. [msi.EnsureFresh()](https://github.com/Azure/go-autorest/blob/10e0b31633f168ce1a329dcbdd0ab9842e533fb5/autorest/adal/token.go#L882) will update the token if it is about to expire, and then msi.OAuthToken() gets the token to be used for the connection

kfcampbell commented:
I like these changes and I'm looking forward to being able to use them in the usage service! Managed identity SQL auth will save us a lot of headaches around rolling passwords and disaster recovery.

Do you have thoughts on the strategy to merge with the upstream repo after this merges in your branch? Some idle thoughts:
- Their [contributing docs](https://github.com/golang-migrate/migrate/blob/3dfb0ff538aaebd744c6b8145ee7581f06c7a9ed/CONTRIBUTING.md) aren't suuuper helpful. 
- Maybe it's best to merge here and then open up a PR in their repo. 
- `@taywrobel` works here and has [committed to the project in the past](https://github.com/golang-migrate/migrate/commits?author=taywrobel). He might be able to help us or tell us what he did to get his CockroachDB changes in.

samfoxcode committed:
Update database/sqlserver/README.md

Co-authored-by: Keegan Campbell <me@kfcampbell.com>

samfoxcode commented:
> I like these changes and I'm looking forward to being able to use them in the usage service! Managed identity SQL auth will save us a lot of headaches around rolling passwords and disaster recovery.
> 
> Do you have thoughts on the strategy to merge with the upstream repo after this merges in your branch? Some idle thoughts:
> 
> * Their [contributing docs](https://github.com/golang-migrate/migrate/blob/3dfb0ff538aaebd744c6b8145ee7581f06c7a9ed/CONTRIBUTING.md) aren't suuuper helpful.
> * Maybe it's best to merge here and then open up a PR in their repo.
> * `@taywrobel` works here and has [committed to the project in the past](https://github.com/golang-migrate/migrate/commits?author=taywrobel). He might be able to help us or tell us what he did to get his CockroachDB changes in.

that was my initial thought too, have it reviewed here first then merge into the fork's main, then open a PR against the upstream

kfcampbell commented:
That sounds good to me! I'm looking forward to seeing this PR on the main module. I won't "approve" this for now since I'm not quite sure of the etiquette in this fork scenario, but I'd definitely be interested in getting the maintainers' feedback soon.

samfoxcode committed:
Update sqlserver.go

refactor resource uri logic into its own method

samfoxcode committed:
Update sqlserver_test.go

add tests for msi connection. can only test whether it fails with useMsi= true, or succeeds with useMsi=false
